packages/expo-device/android/src/main/java/expo/modules/device/DeviceModule.java

```diff
@@ -0,0 +1,385 @@
+    try {
+      BigInteger storage = this.getFreeDiskStorage();
+      String stringValue = storage.toString();
+      promise.resolve(stringValue);
+    } catch (NullPointerException e) {
+      e.printStackTrace();
+      promise.reject(e);
+    }
+  }
+
+  @ExpoMethod
+  public void getIPAddressAsync(Promise promise) {
+    try {
+      Integer ipAddress = getWifiInfo().getIpAddress();
+      // Convert little-endian to big-endianif needed
+      if (ByteOrder.nativeOrder().equals(ByteOrder.LITTLE_ENDIAN)) {
+        ipAddress = Integer.reverseBytes(ipAddress);
+      }
+
+      byte[] ipByteArray = BigInteger.valueOf(ipAddress).toByteArray();
+
+      String ipAddressString;
+      try {
+        ipAddressString = InetAddress.getByAddress(ipByteArray).getHostAddress();
+      } catch (Exception ex) {
+        Log.e("WIFIIP", "Unable to get host address.");
+        ipAddressString = null;
+      }
+      promise.resolve(ipAddressString);
+    } catch (Exception e) {
+      Log.e(TAG, e.getMessage());
+      promise.reject(e);
+    }
+  }
+
+  @ExpoMethod
+  public void getMACAddressAsync(Promise promise) {
+
+    String macAddress = getWifiInfo().getMacAddress();
+
+    String permission = "android.permission.INTERNET";
+    int res = mContext.checkCallingOrSelfPermission(permission);
+
+    if (res == PackageManager.PERMISSION_GRANTED) {
+      try {
+        List<NetworkInterface> all = Collections.list(NetworkInterface.getNetworkInterfaces());
+        for (NetworkInterface nif : all) {
+          if (!nif.getName().equalsIgnoreCase("wlan0")) continue;
+
+          byte[] macBytes = nif.getHardwareAddress();
+          if (macBytes == null) {
+            macAddress = "";
+          } else {
+
+            StringBuilder res1 = new StringBuilder();
+            for (byte b : macBytes) {
+              res1.append(String.format("%02X:", b));
+            }
+
+            if (res1.length() > 0) {
+              res1.deleteCharAt(res1.length() - 1);
+            }
+            macAddress = res1.toString();
+          }
+        }
+      } catch (Exception ex) {
+      }
+    }
+
+    promise.resolve(macAddress);
+  }
+
+  @ExpoMethod
+  public void isAirplaneModeAsync(Promise promise) {
+    boolean isAirPlaneMode;
+
+    isAirPlaneMode = Settings.Global.getInt(mContext.getContentResolver(), Settings.Global.AIRPLANE_MODE_ON, 0) != 0;
+
+    promise.resolve(isAirPlaneMode);
+  }
+
+  @ExpoMethod
+  public void hasSystemFeatureAsync(String feature, Promise promise) {
+    if (feature == null || feature == "") {
+      promise.resolve(false);
+      return;
+    }
+
+    boolean hasFeature = mContext.getApplicationContext().getPackageManager().hasSystemFeature(feature);
+    promise.resolve(hasFeature);
+  }
+
+  @ExpoMethod
+  public void isPinOrFingerprintSetAsync(Promise promise) {
+    KeyguardManager keyguardManager = (KeyguardManager) mContext.getApplicationContext().getSystemService(Context.KEYGUARD_SERVICE); //api 16+
```

Comment:
Do we need to guard this version?

Reply:
This KeyguardManager is to check whether this device is secured by a Pin/Password/Fingerprint. It didn't actually guard anything though.